develop/2017-05-29-fgrehm-analysis-on-meals-based-on-presence-information.py

```diff
@@ -0,0 +1,563 @@
+
+presences = presences.query('presence == "Present"')
+print("Presence records with presence confirmed:", len(presences))
+
+# Match presence with session and deputy info
+presences = pd.merge(presences, sessions, on=['date', 'session'])
+print("Presence records with session matched:", len(presences))
+
+presences = pd.merge(deputies, presences, on='congressperson_document')
+print("Presence records with deputies matched:", len(presences))
+
+
+# In[11]:
+
+presences['first_session_at'] = presences['started_at']
+presences['last_session_at'] = presences['started_at']
+presences['total_sessions'] = 0
+aggregations = {
+    'first_session_at': 'min',
+    'last_session_at': 'max',
+    'total_sessions': 'count',
+    'session': lambda x: "{%s}" % ', '.join(x.astype(str)),
+}
+presences = presences.groupby(['congressperson_id', 'date'], as_index=False).agg(aggregations)
+print("Confirmed presences grouped by date and congressperson:", len(presences))
+
+
+# In[12]:
+
+meals_outside_df = meals.query('company_state != "DF"')
+meals_outside_df_while_in_df = pd.merge(
+    left=presences, 
+    right=meals_outside_df, 
+    left_on=['congressperson_id', 'date'],
+    right_on=['congressperson_id', 'issue_date']
+)
+print("Meals outside DF on days that the congressperson was there:", len(meals_outside_df_while_in_df))
+
+meals_outside_df_while_in_df = meals_outside_df_while_in_df.query('company_name != "GOL LINHAS AEREAS S.A."')
+print("Meals excluding expenses on flight:", len(meals_outside_df_while_in_df))
+
+meals_outside_df_while_in_df = meals_outside_df_while_in_df[
+    ~meals_outside_df_while_in_df.company_name.str.contains('HOTEL')
+]
+# We keep hotels out of the equation for this analysis since it they usually
+# have meals made on multiple days. Not to say that the bill might dated from
+# the day that the congressperson was back.
+print("Meals excluding expenses on hotels:", len(meals_outside_df_while_in_df))
+
+
+# In[13]:
+
+def score(meal):
+    return score_by_time(meal) + score_by_ts(meal) - penalty_by_date(meal)
+
+def score_by_time(meal):
+    if meal.total_sessions > 1:
+        return meal.last_session_at.hour - meal.first_session_at.hour + (meal.total_sessions - 1)
+    else:
+        return 0
+    
+def score_by_ts(meal):
+    if meal.timestamps == '':
+        return 0
+    
+    occurences = 0
+    for ts in meal.timestamps:
+        if ts == None:
+            continue
+        if (meal.first_session_at.hour-1) < ts.tm_hour < (meal.last_session_at.hour+1):
+            occurences += 1
+    return occurences * 10
+
+def penalty_by_date(meal):
+    if meal.receipt_dates == '':
+        return 0
+    
+    bad_occurences = 0
+    for dt in meal.receipt_dates:
+        if dt == None:
+            continue
+        if dt != meal.issue_date:
+            bad_occurences += 1
+    return bad_occurences * 2
+
+def period_in_df(meal):
+    url = (
+        'http://www.camara.leg.br/SitCamaraWS/sessoesreunioes.asmx/ListarPresencasDia?'
```

Comment:
This was added in order to support the manual checks in case anyone is up for making a report to the chamber of deputies or something else. The link it generates points to the data from where we got the presences information. Ideally it should point to a human friendly page instead of an XML but I couldn't think of a way to build that url.

Ex: http://www.camara.leg.br/internet/votacao/mostraPresenca.asp?numLegislatura=55&codCasa=1&numSessaoLegislativa=3&indTipoSessaoLegislativa=O&numSessao=143&indTipoSessao=E&tipo=2&codCasaParlamentar=1 & http://www.camara.leg.br/internet/votacao/mostraPresenca.asp?numLegislatura=55&codCasa=1&numSessaoLegislativa=3&indTipoSessaoLegislativa=O&numSessao=143&indTipoSessao=E&tipo=1&codCasaParlamentar=1

/cc @rodolfo-viana